Remove Argument.AllowedValues, use Validator to implement the logic

adamsitnik opened the pull request:
follow up to https://github.com/dotnet/command-line-api/issues/1891#issuecomment-1310520011

adamsitnik commented on src/System.CommandLine/Parsing/ArgumentResult.cs:
```diff
@@ -83,7 +83,7 @@ public void OnlyTake(int numberOfTokens)
                 if (!string.IsNullOrWhiteSpace(ErrorMessage))
                 {
-                    return new ParseError(ErrorMessage!, this);
+                    return new ParseError(ErrorMessage!, Parent is OptionResult option ? option : this);
```
This change is not obvious. Before this change, the `SymbolResult` of `ParseError` for an Argument with unmatched tokens owned by an Option was the `OptionResult`, not `ArgumentResult`:

https://github.com/dotnet/command-line-api/blob/0adeb26f4ce52c454280d9f3a5d6975e5d081fec/src/System.CommandLine/Parsing/ParseResultVisitor.cs#L509

but this was true only "non-custom errors" (user defined, coming from validators):

https://github.com/dotnet/command-line-api/blob/0adeb26f4ce52c454280d9f3a5d6975e5d081fec/src/System.CommandLine/Parsing/ParseResultVisitor.cs#L510

Without this change, one test was failing:

```log
  Failed System.CommandLine.Tests.ParseDiagramTests.Parse_result_diagram_displays_unmatched_tokens [283 ms]
  Error Message:
   Expected string to be
"[ command ![ -x <ar> ] ]", but
"[ command [ -x !<ar> ] ]" differs near "[ -" (index 10).
```

Since we hide the `Option.Argument` from the user, I think it's reasonable to use the `OptionResult` in such cases for all validations.

jonsequitur commented:
If there's no test for case-insensitive enum parsing, we should definitely add one.

Jozkee commented on src/System.CommandLine/Argument.cs:
```diff
@@ -105,14 +103,15 @@ private protected override string DefaultName
         /// </summary>
         /// <param name="validate">The action to validate the parsed argument.</param>
-        public void AddValidator(Action<ArgumentResult> validate) => Validators.Add(validate);
+        public void AddValidator(Action<ArgumentResult> validate) => (_validators ??= new()).Add(validate);
```
There are other places that call Validator.Add, which would now add the Action to the Array.Empty<Action<ArgumentResult>>.

https://github.com/dotnet/command-line-api/blob/350a618ab44932c568ae2392d7a571281baed59a/src/System.CommandLine/Argument%7BT%7D.cs#L215

https://github.com/dotnet/command-line-api/blob/350a618ab44932c568ae2392d7a571281baed59a/src/System.CommandLine/OptionValidation.cs#L21

adamsitnik commented:
I've addressed all the feedback, I am going to merge this PR now to unblock other PRs.

> AcceptOnlyFromAmong is case-sensitive, unlike the Enum.Parse call in ArgumentConverter

Please create an issue for that. We can just extend `AcceptOnlyFromAmong ` with an `StringComparison comparisonType` argument and let the users decide
